File: src/main/java/com/github/sorusclient/client/module/Module.java
package com.github.sorusclient.client.module;

import com.github.sorusclient.client.setting.Setting;
import com.github.sorusclient.client.setting.SettingContainer;
import com.github.sorusclient.client.setting.Util;
import com.github.sorusclient.client.util.Pair;

import java.util.*;

public class Module implements SettingContainer {

    private final String id;
    private final Map<String, Setting<?>> settings = new HashMap<>();

    protected final Setting<Boolean> isShared;

    public Module(String id) {
        this.id = id;

        this.isShared = new Setting<>(false);
    }

    protected void register(String id, Setting<?> setting) {
        this.settings.put(id, setting);
    }

    @Override
    public String getId() {
        return this.id;
    }

    @Override
    public void load(Map<String, Object> settings) {
        for (Map.Entry<String, Object> setting : settings.entrySet()) {
            Setting<?> setting1 = this.settings.get(setting.getKey());
            if (setting1 != null) {
                setting1.setValueRaw(Util.toJava(setting1.getType(), setting.getValue()));
            }
        }
    }

    @Override
    public Map<String, Object> save() {
        Map<String, Object> settingsMap = new HashMap<>();
        for (Map.Entry<String, Setting<?>> setting : this.settings.entrySet()) {
            settingsMap.put(setting.getKey(), Util.toData(setting.getValue().getValue()));
        }
        return settingsMap;
    }

    public List<Pair<Pair<String, Setting<?>>, Pair<String, Object>>> getSettings() {
        return Collections.emptyList();
    }

    @Override
    public boolean isShared() {
        return this.isShared.getValue();
    }

    @Override
    public void setShared(boolean isShared) {
        this.isShared.setValue(isShared);
    }

}

File: src/main/kotlin/com/github/sorusclient/client/module/Module.kt
package com.github.sorusclient.client.module

import com.github.sorusclient.client.setting.ConfigurableData
import com.github.sorusclient.client.setting.ConfigurableData.Toggle
import com.github.sorusclient.client.setting.Setting
import com.github.sorusclient.client.setting.SettingContainer
import com.github.sorusclient.client.setting.Util

open class Module(override val id: String) : SettingContainer {

    private val settings: MutableMap<String, Setting<*>> = HashMap()
    private val sharedInternal: Setting<Boolean> = Setting(false)

    protected fun register(id: String, setting: Setting<*>) {
        settings[id] = setting
    }

    override fun load(settings: Map<String, Any>) {
        for ((key, value) in settings) {
            val setting1 = this.settings[key]
            Util.toJava(setting1?.type, value)?.let { setting1?.setValueRaw(it) }
        }
    }

    override fun loadForced(settings: Map<String, Any>) {
        for ((key, value) in settings) {
            val setting1 = this.settings[key]!!
            val forcedValues: MutableList<Any> = ArrayList()
            if (value is List<*>) {
                for (element in value) {
                    forcedValues.add(Util.toJava(setting1.type, element)!!)
                }
            } else {
                forcedValues.add(Util.toJava(setting1.type, value)!!)
            }
            setting1.setForcedValueRaw(forcedValues)
        }
    }

    override fun removeForced() {
        for (setting in settings.values) {
            setting.setForcedValueRaw(null)
        }
    }

    override fun save(): Map<String, Any> {
        val settingsMap: MutableMap<String, Any> = HashMap()
        for ((key, value) in settings) {
            settingsMap[key] = Util.toData(value.value!!)
        }
        return settingsMap
    }

    open fun addSettings(settings: MutableList<ConfigurableData>) {
        settings.add(Toggle("Shared", sharedInternal))
    }

    override var shared: Boolean = false

}